experimental/Geometry/Schemes.jl

```diff
@@ -0,0 +1,1244 @@
+import AbstractAlgebra.Ring, Oscar.AlgHom, Oscar.compose, AbstractAlgebra.Generic.Frac
+import Base: ∘
+import Oscar: base_ring
+import AbstractAlgebra: FPModule, FPModuleElem
+
+include( "./Multiindices.jl" )
+using Oscar.Multiindices
+
+include( "./Misc.jl" )
+using Oscar.Misc
+
+export AffineScheme, PrincipalSubScheme,Spec, SpecPrincipalOpen, affine_space
+export AffSchMorphism, base_ring,ambient_ring,defining_ideal, pullback, pullback_from_parent, pullback_from_root, inclusion_in_parent, inclusion_in_root, set_name!
+export localize, get_unit
+
+export AffSchMorphism
+export domain, codomain, pullback
+
+export Glueing
+export first_patch, second_patch, overlap, first_inclusion, second_inclusion, glueing_isomorphism
+
+export Covering
+export get_patches, get_glueings, add_patch
+
+export Refinement
+
+export CoveredScheme
+export get_coverings
+
+export AbstractCoherentSheaf
+
+export IdealSheaf
+export parent, covering, ideals
+
+export CoherentSheaf
+export parent, covering, modules
+
+export LineBundle
+export parent, covering, transitions, OO
+
+abstract type Scheme{ S <: Ring }end
+abstract type AffineScheme{S, T <: MPolyRing, U <: MPolyElem} <: Scheme{S} end
+abstract type SchemeMorphism end
+
+abstract type SchematicPoint end
+abstract type Sheaf end
+abstract type AbstractCoherentSheaf <: Sheaf end
+
+####################################################################################
+# The classical affine scheme, explicitly given as the quotient 
+# ring of a polynomial algebra.
+mutable struct Spec{S,T,U} <: AffineScheme{S,T,U}
+  # the basic fields 
+  k::S			# the base ring (usually a field) of definition for the scheme
+  R::T		  	# the ambient polynomial ring to model this affine scheme
+  I::MPolyIdeal{U}	# The ideal in R defining the scheme
+
+  # fields for caching
+  name::String # the name of this scheme for printing
+
+  function Spec(k::S, R::T, I::MPolyIdeal{U} ) where{
+			S <: Ring, T <:MPolyRing , U <: MPolyElem}
+    if k != coefficient_ring(R)
+      error( "Base ring of the affine scheme does not coincide with the base ring of the associated algebra." )
```

Comment:
No trailing periods in error message
```suggestion
      error("Base ring of the affine scheme does not coincide with the base ring of the associated algebra")
```

Reply:
Ok. Again, good to know.